src/charm.py

```diff
@@ -8,35 +8,208 @@
         super().__init__(*args)
-        self.framework.observe(self.on.install, self._on_install)
+        self.certificates_provider = TLSCertificatesRequiresV2(self, REQUIRES_RELATION_NAME)
+        self.certificates_requirers = TLSCertificatesProvidesV2(self, PROVIDES_RELATION_NAME)
+        self.framework.observe(self.on.install, self._update_status)
+        self.framework.observe(self.on.update_status, self._update_status)
+        self.framework.observe(self.on.certificates_requires_relation_joined, self._update_status)
+        self.framework.observe(self.on.certificates_provides_relation_joined, self._update_status)
+        self.framework.observe(
+            self.certificates_requirers.on.certificate_creation_request,
+            self._on_certificate_creation_request,
+        )
+        self.framework.observe(
+            self.certificates_requirers.on.certificate_revocation_request,
+            self._on_certificate_revocation_request,
+        )
+        self.framework.observe(
+            self.certificates_provider.on.certificate_available,
+            self._on_certificate_available,
+        )
+        self.framework.observe(
+            self.certificates_provider.on.certificate_invalidated,
+            self._on_certificate_invalidated,
+        )
+        self.framework.observe(
+            self.certificates_provider.on.all_certificates_invalidated,
+            self._on_all_certificates_invalidated,
+        )
 
-    def _on_install(self, event: InstallEvent) -> None:
-        """Handles the install event.
+    def _update_status(self, event: EventBase) -> None:
+        """Handles charm events that need to update the status.
 
-        The charm will be in Active Status and ready to handle actions.
+        The charm will be in Active Status when related to a TLS Provider
+        and Blocked status otherwise.
 
         Args:
-            event (InstallEvent): Juju event.
+            event (EventBase): Juju event.
 
         Returns:
             None
         """
+        if not self.model.get_relation(REQUIRES_RELATION_NAME):
+            self.unit.status = BlockedStatus("Need a relation to a TLS certificates provider")
+            return
         self.unit.status = ActiveStatus()
 
+    def _on_certificate_creation_request(self, event: CertificateCreationRequestEvent) -> None:
+        """Handles certificate creation request events.
+
+        If a TLS provider is not integrated to this charm, the event will be
+        defered and the status will be Blocked.
+        Otherwise, the request will be forwarded to the provider.
+
+        Args:
+            event (CertificateCreationRequestEvent): Event containing the request
+
+        Returns:
+            None
+        """
+        if not self.model.get_relation(REQUIRES_RELATION_NAME):
+            event.defer()
+            self.unit.status = BlockedStatus("Need a relation to a TLS certificates provider")
+            return
```

Comment:
I think using defer here makes the code simpler, and defer is not going away anytime soon. I also do not expect this defer to be hit often, as this will be expected to be deployed with a TLS provider.